src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Claims/WindowsClaimSet.cs

```diff
@@ -252,5 +319,42 @@ public static bool TryCreateWindowsSidClaim(WindowsIdentity windowsIdentity, out
             claim = null;
             return false;
         }
+
+        class GroupSidClaimCollection : Collection<Claim>
+        {
+            public GroupSidClaimCollection(ClaimsIdentity claimsIdentity, LdapSettings ldapSettings)
+            {
+                if(claimsIdentity is WindowsIdentity)
+                {
+                    var windowsIdentity = (WindowsIdentity)claimsIdentity;
+                    if (windowsIdentity.Token != IntPtr.Zero)
+                    {
+                        foreach (var groupId in windowsIdentity.Groups)
+                        {
+                            var group = groupId.Translate(typeof(NTAccount));
+                            string[] domainGroups = group.Value.Split(new char[] { '\\' });
+                            if (domainGroups.Length > 1)
+                            {
+                                base.Add(new Claim(ClaimTypes.Role, domainGroups[1], Rights.Identity));
+                            }
+                            else
+                            {
+                                base.Add(new Claim(ClaimTypes.Role, group, Rights.Identity));
+                            }
+                        }
+                    }
+                }
+                else if(ldapSettings !=null)
+                {
+                    List<Claim> allClaims = LdapAdapter.RetrieveClaimsAsync(ldapSettings, claimsIdentity.Name).GetAwaiter().GetResult();
+                    //should we make above sync -- this would be in the Open path..
+                    foreach(Claim roleClaim in allClaims)
```

Comment:
This is a difficult one. Most of the time we don't need to execute this code, whic is why Groups is lazily created. We don't want to do the work and allocate the memory if it's not needed. But as this requires off box work, if needed it would be a good idea to kick off fetching this as soon as possible in anticipation. I see 2 possibilities:  
1. Leave it as is and let it block on async work
2. Initially have this behavior where it blocks on async work, but set a static boolean field recording that this feature was used in the process so one the second usage of this class we pre-emptively kick this off asynchronously and have the result available as a Task so that hopefully it's completed by the time it's actually needed.

Reply:
going with option 1